### PR TITLE
cli: make config.yaml optional

### PR DESCRIPTION
**Pull request by ShraddhaAg**
### Changelog

- [ ] `CHANGELOG.md` is updated with user-facing content relevant to this PR.

### Affected components

- [X] CLI
- [ ] Docs

### Related Issues

Issue: #1558
RFC: #4668

### CODE REVIEW

**Comment by ShraddhaAg** [comment]
I have not added the check, if `--config-file` flag is provided and config version is V1 in the following commands:

1. all `actions` commands: since this command is supported only from V2, the check isn't needed
2. `hasura docs`: N/A
3. `hasura help`: N/A
4. all `plugin` commands: N/A
5. `hasura update cli`: N/A
6. `hasura version`: N/A

Initially, I took the approach to verify if `--config-file` flag is set within `ec.Validate()` (by checking if `ec.ConfigFile` is already set) but that results in failing [integeration tests for `ec.Validate`](https://github.com/hasura/graphql-engine/blob/master/cli/integration_test/cli_test.go#L56) (as `ec.ConfigFile` is already set in the [`init` command test](https://github.com/hasura/graphql-engine/blob/master/cli/integration_test/cli_test.go#L50), run right before [`ec.validate` test](https://github.com/hasura/graphql-engine/blob/master/cli/integration_test/cli_test.go#L56)). So now I'm doing the check in every command where it's applicable. Hope that makes sense.

File: cli/cli.go [hunk 1]
```diff
@@ -584,17 +588,25 @@ func (ec *ExecutionContext) readConfig() error {
...
-		return errors.Wrap(err, "cannot read config from file/env")
+	if ec.ConfigFile != "" {
+		configBase := filepath.Base(ec.ConfigFile)
+		configName := strings.TrimSuffix(configBase, filepath.Ext(configBase))
+		configPath := filepath.Dir(ec.ConfigFile)
+		v.SetDefault("version", "1")
+		v.SetDefault("metadata_directory", "")
+		v.SetConfigName(configName)
+		v.AddConfigPath(configPath)
+		err := v.ReadInConfig()
```
Comment (scriptonist) [inline cli/cli.go:606]: Will it make sense if we use https://godoc.org/github.com/spf13/viper#Viper.SetConfigFile here ?

File: cli/directory.go [hunk 2]
```diff
@@ -26,7 +34,61 @@ func (ec *ExecutionContext) validateDirectory() error {
...
+// is set relative to project flag.
+func (ec *ExecutionContext) validateConfigFile() error {
+	if len(ec.ExecutionDirectory) == 0 {
+		ec.ExecutionDirectory = filepath.Dir(ec.ConfigFile)
+	} else {
+		if !filepath.IsAbs(ec.ConfigFile) {
+			ec.ConfigFile = filepath.Join(ec.ExecutionDirectory, ec.ConfigFile)
+		}
+	}
+	ec.ConfigFile, _ = filepath.Abs(ec.ConfigFile)
```
Comment (scriptonist) [inline cli/directory.go:84]: Can we handle the error here?

File: cli/directory.go [hunk 3]
```diff
@@ -26,7 +34,61 @@ func (ec *ExecutionContext) validateDirectory() error {
 		ec.ExecutionDirectory = cwd
 	}
 
-	ed, err := os.Stat(ec.ExecutionDirectory)
+	err := validateExecutionDirectory(ec.ExecutionDirectory)
+	if err != nil {
+		return err
+	}
+	ec.ExecutionDirectory, _ = filepath.Abs(ec.ExecutionDirectory)
```
Comment (scriptonist) [inline cli/directory.go:41]: Maybe handle the error?

**Comment by scriptonist** [comment]
While we are at it, I was thinking if It'll be a good idea to move all config related things to a separate package, I've vaguely tried this out here (POC -- will not build)
https://github.com/scriptonist/graphql-engine/tree/config-pkg/cli/internal/config

We can be done with all config related logic in the root command: 
https://github.com/scriptonist/graphql-engine/blob/d465a9b0f7ad7cb8b9ea61fec8519ab7dff78325/cli/commands/root.go#L50

I think, doing so can nicely isolate config setup and makes it a bit more testable. But it requires a fair bit of exploration and validation. 

@ShraddhaAg can you think about this and if you find it worthy, spend some time validating it?

File: cli/cli.go (hunk 1 above)
Comment (ShraddhaAg) [reply cli/cli.go:602]: Ah yes, this makes so much more sense. Updated :+1:

File: cli/directory.go (hunk 2 above)
Comment (ShraddhaAg) [reply cli/directory.go]: Done

File: cli/directory.go (hunk 3 above)
Comment (ShraddhaAg) [reply cli/directory.go]: Done

**Comment by Stefmore02** [comment]
Closing this PR as is no longer relevant.

**Pull request closed without merging**